include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -1047,31 +1046,25 @@ irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr
 
 			uint right = instr_getRightJump(instr);
 			ix = choseRight ? right:(ix+1u);
-			weight *= choseRight ? (1.0-w):w;
+			weight /= rcpChoiceProb;
 		}
 		else 
 #endif //OP_BLEND
 #ifdef OP_COATING
 		if (op==OP_COATING) {
-			// ehh this whole thing of managing COATING is bad af
 			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
 			vec3 eta = bsdf_data_decodeIoR(bsdf_data, OP_COATING)[0];
 			vec3 fresnel = irr_glsl_fresnel_dielectric_frontface_only(eta, currInteraction.isotropic.NdotV);
-			float w = dot(fresnel, CIE_XYZ_Luma_Y_coeffs);
-			float rcpProb;
-			bool choseCoat = irr_glsl_partitionRandVariable(w, u.z, rcpProb);
-			//TODO those weight multiplications are probably wrong, not needed or should concern remainder only
-			if (choseCoat)
-			{
-				weight *= (1.0-w);
-				break;
-			}
-			else
-			{
-				weight *= w;
-				//weight *= diffuse fresnel correction
-				ix = ix+1u;
-			}
+			// TODO include thickness_sigma in weight computation
+			vec3 weight = irr_glsl_diffuseFresnelCorrectionFactor(eta, eta*eta) * (vec3(1.0) - fresnel);
+			float w = dot(weight, CIE_XYZ_Luma_Y_coeffs);
+			w = 1.0 - w;
```

Comment:
you sure that's right?

Reply:
best i could come up with without knowing H and L